18-temperature-and-humidity-monitor/firmware/src/main.cpp

```diff
@@ -0,0 +1,423 @@
+                        notecard.logDebugf("Failed to convert %s to positive "
+                            "float for humidity max.\n", valueStr);
+                    }
+                    else {
+                        state.humidityMax = value;
+                        updated = true;
+                        notecard.logDebugf("Humidity max set to %s%%.\n",
+                            valueStr);
+                    }
+                }
+            }
+            else {
+                notecard.logDebug("NULL body in response to env.get request."
+                    "\n");
+            }
+        }
+
+        notecard.deleteResponse(rsp);
+    }
+    else {
+        notecard.logDebug("NULL response to env.get request.\n");
+    }
+
+    return updated;
+}
+
+// Used in the main loop to check for and accept environment variable updates.
+// If any variables are updated, this function sends a note to Notehub to
+// acknowledge the update.
+void updateEnvVars()
+{
+    if (pollEnvVars()) {
+        if (fetchEnvVars()) {
+            J *req = notecard.newRequest("note.add");
+            if (req != NULL) {
+                JAddStringToObject(req, "file", "notify.qo");
+                JAddBoolToObject(req, "sync", true);
+
+                J *body = JCreateObject();
+                if (body != NULL) {
+                    JAddStringToObject(body, "message", "environment variable "
+                        "update received");
+                    JAddItemToObject(req, "body", body);
+                    notecard.sendRequest(req);
+                }
+                else {
+                    notecard.logDebug("Failed to create note body for env var "
+                        "update ack.\n");
+                }
+            }
+            else {
+                notecard.logDebug("Failed to create note.add request for env "
+                    "var update ack.\n");
+            }
+        }
+    }
+}
+
+// One-time Arduino initialization.
+void setup()
+{
+    // Set up debug output via serial connection.
+    delay(2500);
+    Serial.begin(115200);
+    notecard.setDebugOutputStream(Serial);
+
+    // Initialize the physical I/O channel to the Notecard.
+    Wire.begin();
+    notecard.begin();
+
+    if (!bmeSensor.beginI2C()) {
+        notecard.logDebug("bmeSensor.beginI2C() failed.\n");
+    } 
+    else {
+        notecard.logDebug("Connected to BME280.\n");
+    }
+
+    // Set defaults.
+    state.monitorIntervalMs = MONITOR_INTERVAL_DEFAULT_MS;
+    state.temperatureMin = TEMPERATURE_MIN_DEFAULT;
+    state.temperatureMax = TEMPERATURE_MAX_DEFAULT;
+    state.humidityMin = HUMIDITY_MIN_DEFAULT;
+    state.humidityMax = HUMIDITY_MAX_DEFAULT;
+
+    // Configure the productUID and instruct the Notecard to sync outbound notes
+    // a minimum of every OUTBOUND_SYNC_MINS minutes.
+    J *req = notecard.newRequest("hub.set");
+    JAddStringToObject(req, "product", PRODUCT_UID);
```

Comment:
`PRODUCT_UID` may be the empty string, so best to make this conditional

```c++
if (PRODUCT_UID[0]) {
   JAddStringToObject(req, "product", PRODUCT_UID);
}
```
(I noticed I didn't in NF8, so will also fix that one up.)